### ISSUE TITLE
Electrum plugin thoughts

### ISSUE THREAD

**Issue by chris-belcher** [chris-belcher/joinmarket#44]
Other electrum plugins I've seen fit into one python file, this project has about 12 files which seems impractical.

Perhaps the joinmarket application could run as separately and the electrum plugin merely uses some kind of interprocess communication.

**Comment by AdamISZ** [comment]
My first question about Electrum integration is - is the plan to use the Electrum wallet keys themselves (e.g. the HD in 2.0?) I'm guessing not - so in that case you'd be using, say, `sendpayment.py`, would you need first a send from the Electrum addr to the JM addr, and then do the join?

As for the # files, well, yeah, but first I'd need to look into the whole plugin architecture, which I haven't done yet.

**Comment by chris-belcher** [comment]
We have to use the Electrum wallet keys themselves so the coinjoin can be almost instantly negotiated. So in the GUI there is just a button that's clicked and the tx goes out very quickly.

With `sendpayment.py` and requiring first a send, there would be a wait while the tx confirms.

This probably means changing the architecture slightly so the Taker class doesn't need an instance of the Wallet class in its current form. This is partly what I meant by #13. Electrum can tell Joinmarket the coinjoin address, change address, list of utxos and the private keys needed for spending them.

**Comment by AdamISZ** [comment]
>We have to use the Electrum wallet keys themselves so the coinjoin can be almost instantly negotiated. So in the GUI there is just a button that's clicked and the tx goes out very quickly.

The mixing levels concept will need to be enforced within Electrum then. I guess that needs investigating?

**Comment by chris-belcher** [comment]
I dont think so, the mixing depths are only needed when sending coinjoins to your own wallet. People using electrum are just sending to addresses outside their wallet.

Unless people are sending to an address of theirs in the same wallet, in which case we need to warm them not to do that.

**Comment by chris-belcher** [comment]
I looked a little bit around electrum's code. The relevant source is in the wallet.py file probably, in a plugin we add hooks such as make_unsigned_transaction and sign_transaction. We'd get the inputs, outputs and keys required from those. The sign_transaction hook gives us the password and we can use wallet.get_private_key_from_xpubkey() along with tx.inputs_to_sign()

Not too sure about actually doing pushtx()

**Comment by chris-belcher** [comment]
Uploaded some files here https://github.com/chris-belcher/joinmarket/tree/electrum-plugin

**Comment by ABISprotocol** [comment]
cc'ing @neocogent on this
